aiographite/aiographitesend.py

```diff
@@ -0,0 +1,449 @@
+import asyncio
+from aiographite.graphite_escaping import metrics_name_to_graphite, metrics_name_from_graphite
+import pickle
+import struct
+import socket
+import os
+import time
+
+
+DEFAULT_GRAPHITE_PLAINTEXT_PORT = 2003
+DEFAULT_GRAPHITE_PICKLE_PORT = 2004
+SUPPORT_PROTOCOLS = ["plaintext", "pickle"]
+
+
+class AioGraphiteSendException(Exception):
+    pass
+
+
+class AsyncioGraphiteSendService(object):
+
+	def __init__(self, graphite_server, graphite_port, protocol = "pickle"):
+
+		self.graphite_server_address = (graphite_server, graphite_port)
+
+		self.connect_to_graphite()
+
+		self.protocol = protocol
+
+		self.loop = asyncio.get_event_loop()
+
+
+	def connect_to_graphite(self):
+		"""
+			Connect to Graphite Server based on Provided Server Address
+		"""
+		try:
+			self.socket = socket.create_connection(self.graphite_server_address)
+			self.socket.setblocking(False)
+		except socket.gaierror:
+			raise AioGraphiteSendException("Unable to connect to the provided server address %s:%s" % self.graphite_server_address)
+		except Exception as e:
+			raise e
+		return self.socket
+
+
+
+	def send_single_data(self, metric_dir_list, value, timestamp = None):
+		"""
+			@example: 
+				Assuming that 
+
+					Expected_Metric_Name  =  metaccounts.authentication.password.attempted
+
+				Then input metric_dir_list should be
+
+					metric_dir_list = [metaccounts, authentication, password, attempted]
+
+			@metric_dir_list: List of string
+			@timestamp: the type should be int
+
+			If you're very confident that the metric name is valid, then use <method: send_single_valid_data> instead.
+
+		"""		
+		valid_metric_name = self.to_graphite_valid_metric_name(metric_dir_list)
+		self.send_single_valid_data(valid_metric_name, value, timestamp)
+
+
+
+	def send_dataset_list(self, dataset, timestamp = None):
+		"""
+			@param: 
+				Support two kinds of dataset
+
+				1)	dataset = [(metric_dir_list, value1), (metric_dir_list, value2), ...] 
+
+				or 
+
+				2)	dataset = [(metric_dir_list1, value1, timestamp1), (metric_dir_list1, value2, timestamp2), ...]
+
+			If you're very confident that the metric name is valid, then use <method: send_valid_dataset_list> instead.
+
+		"""
+		if not dataset:
+			return 
+
+		if len(dataset[0]) == 2:
+			valid_dataset = [(self.to_graphite_valid_metric_name(metric_dir_list), value) for metric_dir_list, value in dataset]
+		else:
+			valid_dataset = [(self.to_graphite_valid_metric_name(metric_dir_list), value, timestamp) for metric_dir_list, value, timestamp in dataset]
+
+		self.send_valid_dataset_list(valid_dataset, timestamp)
+
+
+
+	def send_single_valid_data(self, metric, value, timestamp = None):
+		"""
+			@metric: String
```

Comment:
for Python 2, we should use sphinx's documentation syntax:

http://www.sphinx-doc.org/en/stable/domains.html#info-field-lists